src/main/java/jenkins/plugins/openstack/compute/JCloudsSlave.java

```diff
@@ -90,9 +93,7 @@ public JCloudsSlave(
                 labelString,
                 null,
                 new JCloudsRetentionStrategy(),
-                Collections.singletonList(new EnvironmentVariablesNodeProperty(
-                        new EnvironmentVariablesNodeProperty.Entry("OPENSTACK_PUBLIC_IP", Openstack.getAccessIpAddress(metadata))
-                ))
+                mkNodeProperties(Openstack.getAccessIpAddress(metadata), slaveOptions.getNodeProperties())
```

Comment:
Old code provided a hard-coded NodeProperty that set the env var OPENSTACK_PUBLIC_IP.
New code takes the configured list of NodeProperties (if any) and adds in the env var, so the node is created with both the user-specified NodeProperties plus the IP address.

Reply:
Re-reviewing this, are we making sure individual slaves will get detached instance of individual properties? IOW, the same property object does not get attached to more than one slave?

Reply:
No, we are not sure of that.
The list of properties is detached.
The individual properties (aside from the environment variables property, which gets special treatment) are copied by reference.

FYI the docker plugin uses copy by reference too ... and that seems to work ok like that.
One would hope that node properties are immutable once created so there should be no problems...

...also, I don't know of a generic means of making a fully independent copy of a node property.
I guess it might be possible to serialise to XML and then de-serialise, but I do not think we can assume they'll all support clone or have a copy constructor.

If it's possible to write a method that takes a NodeProperty and returns a deep copy it then that _would_ be useful ... and that'd be useful for the docker plugin and the vSphere plugin too ... but _I_ don't know how that could be done :-/

Reply:
I've been experimenting...
I've come up with the following code:
```
    @SuppressWarnings("unchecked")
    private static <T> T makeCopy(final T templateNP) {
        final String templateNPAsXml = Jenkins.XSTREAM.toXML(templateNP);
        final Object fromXml = (T) Jenkins.XSTREAM.fromXML(templateNPAsXml);
        final T copyOfTemplateNP = (T) fromXml;
        return copyOfTemplateNP;
    }
```

I could put that code into JCloudsSlave.java and then, where the `mergeNodeProperties` method says
```
            } else {
                b.add(templateNP);
            }
```
we'd change it to be
```
            } else {
                final NodeProperty<? extends Node> copyOfTemplateNP = makeCopy(templateNP);
                b.add(copyOfTemplateNP);
            }
```

...and then fix the corresponding unit tests (which are looking for `sameInstance`) and add another test that ensures that the node properties are still `equalTo(...)` but `not(sameInstance(...))`.

@olivergondza WDYT?

Reply:
...or, in order to avoid making a new String each time, we could do:
```
    @SuppressWarnings("unchecked")
    private static <T> T makeCopy(final T original) {
        class IOStream extends ByteArrayOutputStream {
            public ByteArrayInputStream getCurrentContent() {
                return new ByteArrayInputStream(buf, 0, count);
            }
        }
        final IOStream xml = new IOStream();
        Jenkins.XSTREAM.toXML(original, xml);
        final Object copy = Jenkins.XSTREAM.fromXML(xml.getCurrentContent());
        return (T) copy;
    }
```
I can't tell for sure which would be more efficient in reality - in theory, this new version avoids the needless creation of a big String, which _should_ be quicker...

Alternatively, it might be better/quicker to use the `Reader`/`Writer` equivalents (by having the class extend `CharArrayWriter` and return a `CharArrayReader`).

However the original String method makes for much simpler code, even if it might not be the fastest.

Personally, I think that "easily read code" wins over "theoretically faster"...

Reply:
> Personally, I think that "easily read code" wins over "theoretically faster"...

100% agree.  Let's start with that.

Reply:
Implemented in https://github.com/jenkinsci/openstack-cloud-plugin/pull/270/commits/0b13f412d6251ee6cd92f498cf509f4e986e3520 ... along with a new unit-test to ensure "equalTo and not same instance"